tools/codegen/model.py

```diff
@@ -1496,3 +1509,66 @@ def parse_returns(return_decl: str) -> Tuple[Return, ...]:
     if return_decl[0] == '(' and return_decl[-1] == ')':
         return_decl = return_decl[1:-1]
     return tuple(Return.parse(arg) for arg in return_decl.split(', '))
+
+# A PrecomputedElement consists of the name of such an element and its C++ type.
+# Together with `Precompute` below, it helps capture all information about
+# precomputed elements needed to properly generate code for a kernel that has them.
+@dataclass(frozen=True)
+class PrecomputedElement:
+    # The name of the precomputed element.
+    name: str
+    # The C++ type of the precomputed element.
+    cpp_ty: str
+
+    @staticmethod
+    def parse(src: str) -> 'PrecomputedElement':
+        cpp_ty, name = src.split(" ")
+        r = PrecomputedElement(name=name, cpp_ty=cpp_ty)
+        assert str(r) == src, f"{str(r)} != {src}"
+        return r
+
+    def decl(self) -> str:
+        return f"{self.cpp_ty} {self.name}"
+
+    def __str__(self) -> str:
+        return self.decl()
+
+# A Precompute instance consists of a map from kernel argument name
+# to the list of PrecomputedElement instances that should replace that
+# kernel argument in the impl function.
+@dataclass(frozen=True)
+class Precompute:
+    # A map from kernel argument name -> a list of precomputed
+    # elements that replaces/supersedes it.
+    replace: Dict[str, List[Argument]]
```

Comment:
The syntax change was suggested by your data representation here, which doesn't separately store elements lol